bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+    ConnectedServer {
+        accepted_addr: SocketAddrV6,
+        addr: SocketAddrV6,
+        peer_id: Baseboard,
+    },
+    ConnectedClient {
+        addr: SocketAddrV6,
+        peer_id: Baseboard,
+    },
+    Disconnected {
+        peer_id: Baseboard,
+    },
+    Received {
+        from: Baseboard,
+        msg: FsmMsg,
+    },
+    FailedServerHandshake {
+        addr: SocketAddrV6,
+    },
+}
+
+/// Messages sent from the main task to the connection managing tasks
+#[derive(Debug, PartialEq)]
+pub enum MainToConnMsg {
+    Close,
+    Msg(Msg),
+}
+
+// A handle to a task managing a connection to a peer
+//
+// This is either a negotiating (connecting client but not yet handshake
+// complete), or established connection (handshake complete - whether connecting
+// client or accepting server).
+pub struct PeerConnHandle {
+    pub handle: JoinHandle<()>,
+    pub tx: mpsc::Sender<MainToConnMsg>,
+    pub addr: SocketAddrV6,
+    // This is used to differentiate stale `ConnToMainMsg`s from cancelled tasks
+    // with the same addr from each other
+    pub unique_id: u64,
+}
+
+// A handle to a task of an accepted socket, pre-handshake
+pub struct AcceptedConnHandle {
+    pub handle: JoinHandle<()>,
+    pub tx: mpsc::Sender<MainToConnMsg>,
+    pub addr: SocketAddrV6,
+    // This is used to differentiate stale `ConnToMainMsg`s from cancelled tasks
+    // with the same addr from each other
+    pub unique_id: u64,
+}
+
+// Established connection management code running in its own task
+struct EstablishedConn {
+    peer_id: Baseboard,
+    unique_id: u64,
+    write_sock: OwnedWriteHalf,
+    read_sock: OwnedReadHalf,
+    main_tx: mpsc::Sender<ConnToMainMsg>,
+    rx: mpsc::Receiver<MainToConnMsg>,
+    log: Logger,
+    read_buf: Vec<u8>,
+    total_read: usize,
+
+    // Used for managing inactivity timeouts for the coonnection
+    last_received_msg: Instant,
+
+    // Keep a queue to write serialized messages into We limit the queue
+    // size, and if it gets exceeded it means the peer at the other
+    // end isn't pulling data out fast enough. This should be basically
+    // impossible to hit given the size and rate of message exchange
+    // between peers. We go ahead and close the connection if the queue
+    // fills.
+    write_queue: VecDeque<Vec<u8>>,
+
+    // The current serialized message being written if there is one
+    current_write: Cursor<Vec<u8>>,
+}
+
+impl EstablishedConn {
+    fn new(
+        peer_id: Baseboard,
+        unique_id: u64,
+        write_sock: OwnedWriteHalf,
+        read_sock: OwnedReadHalf,
+        main_tx: mpsc::Sender<ConnToMainMsg>,
+        rx: mpsc::Receiver<MainToConnMsg>,
+        log: Logger,
+    ) -> EstablishedConn {
+        EstablishedConn {
+            peer_id,
+            unique_id,
+            write_sock,
+            read_sock,
+            main_tx,
+            rx,
+            log,
+            read_buf: vec![0u8; CONN_BUF_SIZE],
```

Comment:
Should this be a array instead of a `Vec`, since it's always of known size?

(or a [`Box<[u8]>`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from-6) if you want it to be heap-allocated)

Reply:
Good call.

Reply:
fixed in fd92a4bc5